Whiteliste configs for set/unset commands.

hjafarpour opened the pull request:
This patch changes the behavior for SET/UNSET command so only the properties in a predefined whitelist can be configured using these commands and the command would fail user tries to config any other property.
Currently user can run these command with many properties that should not be configurable by users. For instance, user should not be able to issue the following command:

```
ksql> set 'ksql.service.id` = 'foo';
```
Currently we have only 4 configurable session properties, but we can add more in this PR if reviewers have convincing suggestion! :) 
So as of now the properties that user can set/unset from client are the following:

- ksql.sink.window.change.log.additional.retention

- ksql.streams.commit.interval.ms

- auto.offset.reset  

- group.id  

Any other property that is used in SET/UNSET command or is sent to the REST API as part of properties will result in failure of the command/request.  

Since we store all the config properties for each query along with that query, this won't have any effect on the existing queries and therefore no backward compatibility issue in the interactive mode.
In the headless mode, user can easily update the query file by removing invalid set/unset commands.


### Testing done 
Related tests were updated.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

agavra commented on docs/developer-guide/syntax-reference.rst:
```diff
@@ -1124,6 +1124,34 @@ SHOW PROPERTIES
+| ksql.streams.commit.interval.ms                   | The frequency with which to save the position (offsets in source topics) of tasks.         |
+|                                                   | The default is 30000 milliseconds (at-least-once) / 100 milliseconds (exactly-once).       |
++---------------------------------------------------+--------------------------------------------------------------------------------------------+
+| auto.offset.reset                                 | Configure the KSQL queries to read the source topics from earliest or latest offset.       |
```
recently I suggested that someone us `group.id` so that they can get around a bug where PRINT TOPIC uses an empty groupId that doesn't work with ACLs (fyi @big-andy-coates)

JimGalasyn replied:
We can set `auto.offset.reset` in earlier version, can't we?

hjafarpour replied:
Hmm, we can add `group.id` too then since we have a valid reason!
Yes, we always could set `auto.offset.reset`.

agavra commented on docs/developer-guide/syntax-reference.rst:
```diff
@@ -1124,6 +1124,34 @@ SHOW PROPERTIES
+
+.. code:: sql
+
+    SET 'auto.offset.reset' = 'earliest';
```
```suggestion
    [SET/UNSET] 'property_name' = 'property_value';
```
If we want an example, then we can have that below (this follows the rest of the syntax doc)

JimGalasyn replied:
+1 to the example idea.

hjafarpour replied:
Made the change.

JimGalasyn commented on docs/developer-guide/syntax-reference.rst:
```diff
@@ -1124,6 +1124,34 @@ SHOW PROPERTIES
+|                                                   | The default is 30000 milliseconds (at-least-once) / 100 milliseconds (exactly-once).       |
++---------------------------------------------------+--------------------------------------------------------------------------------------------+
+| auto.offset.reset                                 | Configure the KSQL queries to read the source topics from earliest or latest offset.       |
+|                                                   | The default in KSQL is 'latest'.                                                           |
```
\`\`latest\`\` (double-ticks)

vcrfxia commented on docs/developer-guide/syntax-reference.rst:
`````diff
@@ -1124,6 +1124,38 @@ SHOW PROPERTIES
+|                                                   | The default in KSQL is ``latest``.                                                         |
++---------------------------------------------------+--------------------------------------------------------------------------------------------+
+| group.id                                          | A unique string that identifies the consumer group this consumer belongs to.               |
+|                                                   | The default in KSQL is ````.                                                               |
`````
Is the default group ID really empty? I thought it was a something like `_confluent-ksql-<ksql.service.id>query_<query-id>...`? Actually on second thought, this default is probably only referring to the group we use for `PRINT TOPIC`. Should probably clarify that?

hjafarpour replied:
Made it more clear in the description that this is only for PRINT TOPIC command.

hjafarpour commented:
Thanks for the comments @vcrfxia , the main reason to use whitelist instead of blacklist (`IMMUTABLE_PROPERTIES`) is to have more control on the properties that can be configured by users. With whitelist we eliminate probability of unintentional property configuration and we explicitly know what properties are configurable via REST API. 
On the backward compatibility issue, note that the existing queries already have the properties accompanied with them and there won't be any changes for those queries. This only applies to the session properties in the client so if the client already set some properties they will be valid until the end of the current session. When the session ends, the properties expire anyway and user has to set them again which then this PR will enforce the limitation. So all new queries will only have permissible config properties.

vcrfxia commented:
@hjafarpour yes, I was imprecise in my usage of "backwards-incompatible," for which I apologize. You are correct that any existing queries will be unaffected by this PR, and the PR is therefore backwards-compatible. The concern I raised was instead around removing existing functionality. As we discussed offline, I am OK with removing this functionality (specifically, the current ability for users to set many different properties via set/unset) but I think we definitely need to call it out in the release and/or upgrade notes. Probably good to add those changes into this PR so we don't forget later.

Regarding switching from a blacklist to a whitelist, I am again OK with this switch, but would prefer consistency. This PR introduces a whitelist, but also leaves the existing blacklist (in `KsqlEngineProps`) intact, and I worry it will be confusing to have both. Also, placing the whitelist/blacklist in KsqlEngineProps is a stronger check than limiting checks to LocalPropertyValidator as this PR does, since the current blacklist in KsqlEngineProps throws not only on set/unset, but also if a user submits a request (through the REST API) with custom streamsProperties that overrides one of the configs we don't wish to allow users to override. Perhaps consider updating this PR to follow the pattern of the existing blacklist?

hjafarpour commented:
@vcrfxia I updated the PR description to call the change out as you suggested. We should definitely include it in the release notes.
As for the REST API side, I added the whitelist validation to the server side too so now if re receive requests with invalid properties directly in the REST endpoint we validate them. I also added validation for the multiline requests where you can have SET/UNSET commands.
As for the blacklist in `KsqlEngineProps`, I'll keep for now to ensure backward compatibility as we discussed offline.

hjafarpour commented on docs/developer-guide/syntax-reference.rst:
```diff
@@ -1124,6 +1124,39 @@ SHOW PROPERTIES
+
+.. code:: sql
+
+    [SET/UNSET] 'property_name' = 'property_value';
```
That's right! Fixed it :)
